### ISSUE TITLE
Wrong block scoping of "let" in for loop

### ISSUE THREAD

**Issue by ludekstepan** [babel/babel#1078]
Please see gist: https://gist.github.com/ludekstepan/8f7799cd8be8600266ee

**Comment by monsanto** [comment]
That gist is a bit hard to follow. Let me distill that a bit:

```
for (let i = 0; i < 2; i++) {
    () => { i };
    console.log(i);
    i += 1;
}
```

The expected output is `0`, the actual output is `0 1`. The issue is that closing over the loop variable forces Babel to create an IIFE:

```
for (var i = 0; i < 2; i++) {
    (function (i) {
        (function () {
            i;
        });
        console.log(i);
        i += 1;
    })(i);
}
```

... which causes any modifications to `i` to be local to the IIFE.

The best solution I see is to alias the inner `i`, and change any assignments to `i` to also assign to the outer `i`:

```
for (var i = 0; i < 2; i++) {
    (function (_i) {
        (function () {
            _i;
        });
        console.log(_i);
        _i += 1;
        i = _i;
    })(i);
}
```

**Comment by monsanto** [comment]
That gist is a bit hard to follow. Let me distill that a bit:

```
for (let i = 0; i < 2; i++) {
    () => { i };
    console.log(i);
    i += 1;
}
```

The expected output is `0`, the actual output is `0 1`. The issue is that closing over the loop variable forces Babel to create an IIFE:

```
for (var i = 0; i < 2; i++) {
    (function (i) {
        (function () {
            i;
        });
        console.log(i);
        i += 1;
    })(i);
}
```

... which causes any modifications to `i` to be local to the IIFE.

The best solution I see is to alias the inner `i`, and change any assignments to `i` to also assign to the outer `i`:

```
for (var i = 0; i < 2; i++) {
    (function (_i) {
        (function () {
            _i;
        });
        console.log(_i);
        _i += 1;
        i = _i;
    })(i);
}
```

**Comment by ludekstepan** [comment]
@monsanto , thank's for elaborating! However, this is only one part of the problem. When you create a closure inside a loop that accesses the loop variable and execute that closure after the loop has finished, what is the expected outcome?

```javascript
var callback;
for (let i = 0; i < 1; i++) {
  callback = () => console.log(i);
}
callback();
```
Node.js gives '0', Firefox gives '1'

**Comment by monsanto** [comment]
@ludkstepan 0 is the correct answer.

**Comment by ludekstepan** [comment]
@monsanto I would love the 0 being a correct answer as it would be a preferred behavior in most circumstances. But, is there any specification?

**Comment by sebmck** [comment]
@ludekstepan Yes there is. `let` in Firefox has been [notoriously inconsistent](https://bugzilla.mozilla.org/show_bug.cgi?id=950547) due to the non-standard implementation.

**Comment by ludekstepan** [comment]
great, so @monsanto 's solution might work perfectly!

**Comment by sebmck** [comment]
Although no, I don't think the behaviour you're expecting is correct. `let` variables inside loops have a unique binding per iteration so my bets are on the current behaviour being the correct one. I'll have to do some more investigation of the spec to verify the correct behaviour.

**Comment by monsanto** [comment]
@sebmck either way, there is a bug, since without the IIFE it works as @ludkstepan expects

My understanding of the spec is that after a turn of the loop, the current values of each of the per-iteration bindings are used to seed the next per-iteration bindings, the "increment statement" executes, and we repeat the loop. https://people.mozilla.org/~jorendorff/es6-draft.html#sec-forbodyevaluation

For what it's worth this behavior agrees with io.js, which considers let bindings to be stable enough to have on by default.

**Comment by sebmck** [comment]
Thanks for the reference @monsanto! Fixed in the `experimental` branch which will be released as 5.0.0 very soon.
